---
title: "Selective sweep test expected heterozygosity"
output:
  html_document:
    toc: true
    toc_depth: 3   # Adjust depth as needed
date: "2024-09-03"
editor_options: 
  markdown: 
    wrap: 72
---
# 0: Preparation
## Defining the input and output directories
```{r}
# Clean the working environment
rm(list = ls())
knitr::opts_chunk$set(echo = TRUE)
####################################  
# Defining Input parameters
#################################### 
window_size <- 100 * 10^3 # Define the window size for the binning of genomic regions
H_e_decimals_plots <- 3
hotspot_prefix <- "Hotspot_chr"
use_MAF_pruning <- Sys.getenv("use_MAF_pruning")
use_MAF_pruning
min_MAF <- as.numeric(Sys.getenv("min_MAF"))
min_MAF
empirical_breed <- Sys.getenv("empirical_breed")
empirical_breed

####################################  
# Defining the input files
#################################### 
sim_scenario_id <- Sys.getenv("sim_scenario_id")
sim_scenario_id
sweep_test_type <- Sys.getenv("sweep_test_type")
sweep_test_type
####################################  
# Defining the Input Directories
#################################### 
ROH_hotspots_dir <- Sys.getenv("ROH_hotspots_dir")
ROH_hotspots_dir
ROH_hotspot_allele_freq_dir <- Sys.getenv("empirical_roh_hotspots_allele_frequency_dir")
ROH_hotspot_allele_freq_dir
empirical_genomewide_allele_freq_dir <-  Sys.getenv("empirical_allele_frequency_dir")
empirical_genomewide_allele_freq_dir
simulated_model_allele_freq_dir <- Sys.getenv("simulated_model_allele_frequency_dir")
simulated_model_allele_freq_dir

####################################  
# Defining Output Directories
#################################### 
output_dir_sweep_test <- Sys.getenv("output_dir_sweep_test")
output_dir_sweep_test
output_empirical_H_e_dir <- Sys.getenv("output_empirical_H_e_dir")
output_empirical_H_e_dir
output_simulated_model_H_e_dir <- Sys.getenv("output_simulated_model_H_e_dir")
output_simulated_model_H_e_dir

Sys.getenv()  
# Set the working directory for notebook chunks
knitr::opts_knit$set(root.dir = output_dir_sweep_test)
```
## Loading libraries
```{r library()}
library(knitr)
library(ggplot2)
```
# Expected heterozygosity
Takes a list of H_e for a window and computes H_e for each SNP individually
```{r}
calculate_expected_heterozygosity <- function(p) {
  # Calculate expected heterozygosity
  heterozygosity <- 2 * p * (1 - p)
  return(heterozygosity)
}
```
# 1: Loading Hotspot windows
```{r Loading Hotspot windows}
setwd(ROH_hotspots_dir)
# Pattern for finding files containing "causative_variant_window" and ending with ".bed"
pattern <- ".*ROH_Hotspot_windows.*\\.bed$"
hotspot_window_bed_files <- list.files(path = ROH_hotspots_dir, pattern = pattern)
chr_number_pattern <- ".*CHR(\\d+)_.*"
# Extract the chromosome number and window number from the file names
chr_numbers <- gsub(tolower(chr_number_pattern), "\\1", tolower(hotspot_window_bed_files))
# Order the bed_files vector by chromosome number and then by window number
ordered_hotspot_window_bed_files <- hotspot_window_bed_files[order(chr_numbers)]
# Create an empty list to store information for the different ROH-hotspots in
empirical_hotspot_tables <- list()

# Loop through each .bed file (ROH-hotspot allele frequency window-file)
for (file in ordered_hotspot_window_bed_files) {
  # Extract chromosome number and window number from file name
  chromosome <-gsub(tolower(chr_number_pattern), "\\1", tolower(file))
  window <- sub(tolower(".*windows_([0-9]+)_.*"), "\\1", tolower(file))
  
  column_names <- c("CHR","POS1","POS2")
  # Read the .bed file into a data frame, skipping commented lines
  bed_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  window_counter <- 0
  
  # Loop through each row in bed_data
  for (i in 1:nrow(bed_data)) {
    window_counter <- window_counter + 1
    # Extract row data
    row_data <- bed_data[i, ]
    # Create hotspot name
    hotspot_name <- paste(hotspot_prefix, chromosome, "_window_", window_counter, sep = "")
    # Check if the hotspot_name already exists in the list
    if (!(hotspot_name %in% names(empirical_hotspot_tables))) {
      # If it doesn't exist, create a list for it
      empirical_hotspot_tables[[hotspot_name]] <- list()
    }
    
    Start_POS <- row_data$POS1
    End_POS <- row_data$POS2
    # Extract chromosome, start position, and end position
    chr <- chromosome
    start_pos <- Start_POS
    end_pos <- End_POS
    # Calculate the number of 100kbp windows
    num_windows <- ceiling((end_pos - start_pos + 1) / 1e5)
    # Create an empty data frame to store window information
    window_df <- data.frame(CHR = numeric(), Start_POS = numeric(), End_POS = numeric())
    
    # Generate 100kbp windows
    for (j in 1:num_windows) {
      # Calculate start and end positions for each window
      window_start <- start_pos + (j - 1) * 1e5
      window_end <- min(start_pos + j * 1e5 - 1, end_pos)
      # Create a data frame for the current window
      window_data <- data.frame(CHR = chr, POS1 = window_start, POS2 = window_end)
      # Append window_data to window_df
      window_df <- rbind(window_df, window_data)
    }    
    # Create a list with table name, filename, and corresponding data frame
    table_info <- list(CHR = chromosome,Start_POS = Start_POS,End_POS = End_POS, data = window_df)
    # Append the table info to the list
    empirical_hotspot_tables[[hotspot_name]] <- c(empirical_hotspot_tables[[hotspot_name]], table_info)
  }
      }
# View(empirical_hotspot_tables)
```
# 2: Loading the frequency files
## 2.1 Empirical data
### 2.1.1: Empirical ROH-Hotspots
```{r extracting empirical ROH-hotspots}
# Set the working directory to the directory with allele frequencies for the different ROH-hotspots
setwd(ROH_hotspot_allele_freq_dir)

# Get a list of all .bed files in the directory
pattern <- "\\.bed$"
bed_files <- list.files(path = ROH_hotspot_allele_freq_dir, pattern = pattern)

# bed_files <- list.files(pattern = "\\.bed$")
# Extract the chromosome number and window number from the file names
chr_numbers <- as.numeric(gsub(tolower(chr_number_pattern), "\\1", tolower(bed_files)))
# chr_numbers <- as.numeric(sub(".*",chr_prefix,"(\\d+)_.*", "\\1", bed_files))
window_numbers <- as.numeric(sub(tolower(".*windows_(\\d+)_.*"), "\\1", tolower(bed_files)))
# Order the bed_files vector by chromosome number and then by window number
ordered_bed_files <- bed_files[order(chr_numbers, window_numbers)]
# Create an empty list to store information for the different ROH-hotspots in
empirical_hotspot_allele_Freq_tables <- list()
# Initialize a counter for windows without markers
no_marker_count <- 0

# Loop through each .bed file (ROH-hotspot allele frequency window-file)
for (file in ordered_bed_files) {
  # Extract chromosome number and window number from file name
  chromosome <- gsub(tolower(chr_number_pattern), "\\1", tolower(file))
  window <- sub(tolower(".*windows_([0-9]+)_.*"), "\\1", tolower(file))
  # Create table name
  table_name <- paste(hotspot_prefix, chromosome, "_window_", window, "_allele_freq", sep = "")
  # Create table name
  hotspot_name <- paste(hotspot_prefix, chromosome, "_window_", window, sep = "")
  # Read the header line
  con <- file(file, "r")
  header <- readLines(con, n = 1)
  close(con)
  # Remove "#" from the header and split it into column names
  column_names <- sub("#", "", header)
  column_names <- strsplit(column_names, "\t")[[1]]
  # Read the .bed file into a data frame, skipping commented lines
  hotspot_allele_freq_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  
  # Check if pruning markers by MAF should be done
  if (use_MAF_pruning == TRUE) {
    
    # Remove rows where MAF < min_MAF
    hotspot_allele_freq_data <- hotspot_allele_freq_data[hotspot_allele_freq_data$MAF >= min_MAF, ]
  }
  hotspot_allele_freq_data$H_e <- sapply(hotspot_allele_freq_data$MAF, calculate_expected_heterozygosity)

  # Reading in the relevant causative window from  causative_variant_window_tables_Selection_Model
  hotspot_window_data <- empirical_hotspot_tables[[hotspot_name]][["data"]]
  # Initialize a new column to store MAF lists
  hotspot_window_data$MAF_list <- vector("list", nrow(hotspot_window_data))
  # Loop through each 100kbp window for the causative window
  for (j in seq_len(nrow(hotspot_window_data))) {
    genomic_window_start <- hotspot_window_data[j, "POS1"]
    genomic_window_end <- hotspot_window_data[j, "POS2"]
    H_e_list_current_100kbp_window <- c()
    # Subset markers within the current genomic window
    markers_in_window <- hotspot_allele_freq_data[
      hotspot_allele_freq_data$POS1 >= genomic_window_start & hotspot_allele_freq_data$POS1 <= genomic_window_end,
      ]
    if (nrow(markers_in_window) > 0) {
      # Computed the average expected heterozygosity for the current window
      hotspot_window_data[j, "H_e"] <- mean(markers_in_window$H_e)
      # Store the list of MAF values in the new column
      hotspot_window_data$MAF_list[[j]] <- markers_in_window$MAF

    } else {
      # Increment the counter for windows without markers
      no_marker_count <- no_marker_count + 1
      # Skip this window by setting its H_e to NA
      hotspot_window_data[j, "H_e"] <- NA
      hotspot_window_data$MAF_list[[j]] <- list()
    }
  }
   # Calculate and store the average heterozygosity from the 100kbp H_e distribution of the current hotspot
  empirical_hotspot_tables[[hotspot_name]][["data"]] <- hotspot_window_data
  empirical_hotspot_tables[[hotspot_name]]$Avg_He <- mean(hotspot_window_data$H_e, na.rm = TRUE)
  # Check if the selection coefficient already exists in the list
  if (!(hotspot_name %in% names(empirical_hotspot_allele_Freq_tables))) {
    # If it doesn't exist, create a list for it
    empirical_hotspot_allele_Freq_tables[[hotspot_name]] <- list()
  }
  # Create a list with table name and corresponding data frame
  table_info <- list(name = table_name,filename=file, data = hotspot_allele_freq_data)
  # Append the table info to the list
  empirical_hotspot_allele_Freq_tables[[hotspot_name]] <- c(empirical_hotspot_allele_Freq_tables[[hotspot_name]], table_info)
}

# View(empirical_hotspot_allele_Freq_tables)

# markers_after_MAF_pruning <- nrow(empirical_hotspot_allele_Freq_tables[["Hotspot_chr3_window_1"]][["data"]]["H_e"])
# cat("markers after_pruning:",markers_after_MAF_pruning,"\n")
# 
# markers_with_he_01_or_less <- length(empirical_hotspot_allele_Freq_tables[["Hotspot_chr3_window_1"]][["data"]][["H_e"]][empirical_hotspot_allele_Freq_tables[["Hotspot_chr3_window_1"]][["data"]][["H_e"]] < 0.1])
# cat("% markers with H_e <= 0.1:\n", round((markers_with_he_01_or_less/markers_after_MAF_pruning)*100 , 1), "\n")
# 
# fixated_markers <- length(empirical_hotspot_allele_Freq_tables[["Hotspot_chr3_window_1"]][["data"]][["H_e"]][empirical_hotspot_allele_Freq_tables[["Hotspot_chr3_window_1"]][["data"]][["H_e"]] == 0])
# 
# cat("% fixated markers with H_e = 0:\n", round((fixated_markers/markers_after_MAF_pruning)*100 , 1), "\n")
```
### 2.1.2: Empirical - Genomewide
```{r}
# Set the working directory to the directory with allele frequencies for the different ROH-genomewides
setwd(empirical_genomewide_allele_freq_dir)
# Get a list of all .bed files in the directory
bed_files <- list.files(pattern = "\\.bed$")
# bed_files
# Create an empty list to store information for the different ROH-genomewides in
empirical_genomewide_tables <- list()

# Loop through each .bed file (ROH-genomewide allele frequency window-file)
for (file in bed_files) {
  # Extract chromosome number and window number from file name
  chromosome <- sub(".*chr([0-9]+)_.*", "\\1", file)
  print(file)
  print(chromosome)
  window <- sub(".*window_([0-9]+)_.*", "\\1", file)
  # Create table name
  table_name <- paste("Hotspot_chr", chromosome, "_window_", window, "_allele_freq", sep = "")
  # Read the header line
  con <- file(file, "r")
  header <- readLines(con, n = 1)
  close(con)
  
  # Remove "#" from the header and split it into column names
  column_names <- sub("#", "", header)
  column_names <- strsplit(column_names, "\t")[[1]]
  
  # Read the .bed file into a data frame, skipping commented lines
  empirical_genomewide_allele_freq_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  
  # Check if pruning markers by MAF should be done
  if (use_MAF_pruning == TRUE) {
    # Remove rows where MAF < min_MAF
    empirical_genomewide_allele_freq_data <- empirical_genomewide_allele_freq_data[empirical_genomewide_allele_freq_data$MAF >= min_MAF, ]
  }
  empirical_genomewide_allele_freq_data$H_e <- sapply(empirical_genomewide_allele_freq_data$MAF, calculate_expected_heterozygosity)
  table_info <- list(filename = file, data = empirical_genomewide_allele_freq_data)
  # Append the table info to the list
  empirical_genomewide_tables <- c(empirical_genomewide_tables, list(table_info))
}
# View(empirical_genomewide_tables)
# empirical_genomewide_tables[[1]]
```
## 2.2: Simulated data (Simulated model)
```{r}
# Set the working directory to the directory containing the frequency files for the simulated data
setwd(simulated_model_allele_freq_dir)
# Create a pattern for finding all the allele frequency files associated with the studied simulation scenario
pattern <- paste0(".*", sim_scenario_id, "_allele_freq\\.bed$")
# Find all .bed files following the specified pattern
frq_bed_files <- list.files(pattern = pattern)
# Extract the simulation number after "sim_"
simulation_numbers <- as.integer(sub(tolower("^sim_(\\d+)_.*"), "\\1", tolower(frq_bed_files)))
# Sort the simulation-files based on the simulation number
sorted_files <- frq_bed_files[order(simulation_numbers)]
# Initialize an empty list to store Simulated model tables
simulated_model_allele_freq_tables <- list()

# Loop through each .frq file
for (file in sorted_files) {
  # Extract the table name from the file name
  table_name <- gsub(tolower(".frq$"), "",tolower(file))
  column_names <- c("CHR","POS1","POS2","SNP","A1","A2","MAF","NCHROBS")
  allele_freq_data <- read.table(file, header = TRUE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  # Check if pruning markers by MAF should be done
  if (use_MAF_pruning == TRUE) {
    # Remove rows where MAF < min_MAF
    allele_freq_data <- allele_freq_data[allele_freq_data$MAF >= min_MAF, ]
  }
  # Calculate expected heterozygosity for all the markers 
  allele_freq_data$H_e <- sapply(allele_freq_data$MAF, calculate_expected_heterozygosity)
  # Create a list with table name and corresponding data frame
  table_info <- list(name = table_name, data = allele_freq_data)
  # Append the table info to the list
  simulated_model_allele_freq_tables <- c(simulated_model_allele_freq_tables, list(table_info))
}

# sim_id_to_follow <- 1
# 
# markers_after_MAF_pruning <- nrow(simulated_model_allele_freq_tables[[sim_id_to_follow]][["data"]]["H_e"])
# cat("markers after pruning for simulation ",sim_id_to_follow, ": ",markers_after_MAF_pruning,"\n")
# 
# markers_with_he_01_or_less <- length(simulated_model_allele_freq_tables[[sim_id_to_follow]][["data"]]["H_e"][simulated_model_allele_freq_tables[[sim_id_to_follow]][["data"]]["H_e"] < 0.1])
# cat("% markers with H_e <= 0.1:\n", round((markers_with_he_01_or_less/markers_after_MAF_pruning)*100 , 1), "\n")
# 
# fixated_markers <- length(simulated_model_allele_freq_tables[[sim_id_to_follow]][["data"]]["H_e"][simulated_model_allele_freq_tables[[sim_id_to_follow]][["data"]]["H_e"] == 0])
# 
# cat("% fixated markers with H_e = 0:\n", round((fixated_markers/markers_after_MAF_pruning)*100 , 1), "\n")
# View(simulated_model_allele_freq_tables)
```
# 3: Computing expected heterozygosity
Compute H_e for each SNP individually
## 3.1: Empirical data - H_E
### 3.1.1 Genomewide
```{r}
# for (i in seq_along(empirical_genomewide_tables)) {
#   # Extract table name and data frame
#   table_info <- empirical_genomewide_tables[[i]]
#   table_data <- table_info$data
#   
#   # Calculate expected heterozygosity for each window
#   window_H_e_list <- c()  # Initialize an empty list to store expected heterozygosity values
#   
#   # Loop through each row in the table data
#   for (j in 1:nrow(table_data)) {
#     # Extract the MAF (p value)
#     p <- table_data[j, "MAF"]
#     
#     # Calculate expected heterozygosity using the provided function
#     expected_heterozygosity <- calculate_expected_heterozygosity(p)
#     
#     # Append the calculated expected heterozygosity to the list
#     window_H_e_list <- c(window_H_e_list, expected_heterozygosity)
#   }
#   
#   # Assign the calculated expected heterozygosity list to the H_e attribute of table_info
#   empirical_genomewide_tables[[i]]$H_e_list <- window_H_e_list
#   empirical_genomewide_tables[[i]]$Marker_based_Average_H_e <- mean(empirical_genomewide_tables[[i]]$H_e_list)
# }

# View(empirical_genomewide_tables)
```
## 3.2: H_e Simulated data (Simulated model)
```{r}
# # create a vector to store the counts of pruned entries for each simulation
# pruned_counts <- numeric(length(simulated_model_allele_freq_tables))
# datapoints_before_MAF_filtering <- numeric(length(simulated_model_allele_freq_tables))
# sum(simulated_model_allele_freq_tables[[1]][["data"]][["MAF"]] <= min_MAF)
# # sum(simulated_model_allele_freq_tables[[20]][["data"]][["MAF"]] < min_MAF)
# 
# 
# for (i in seq_along(simulated_model_allele_freq_tables)) {
#   
#   
#   # Extract table name and data frame
#   table_info <- simulated_model_allele_freq_tables[[i]]
#   table_data <- table_info$data
#   datapoints_before_MAF_filtering[[i]] <- length(table_data$MAF)
  
#   # Calculate expected heterozygosity for each window
#   window_H_e_list <- c()  # Initialize an empty list to store expected heterozygosity values
#   
#   #create a vector to store the indices of rows to be removed based on having MAF < 0.05 threshold
#   rows_to_remove <- c()
#   
#   # Looping through each row in the table data
#   for (j in 1:nrow(table_data)) {
#     # Extract the MAF (p value)
#     p <- table_data[j, "MAF"]
#     
#     # Calculate expected heterozygosity using the provided function
#     expected_heterozygosity <- calculate_expected_heterozygosity(p)
#     
#     # Append the calculated expected heterozygosity to the list
#     window_H_e_list <- c(window_H_e_list, expected_heterozygosity)
#     
#     if (use_MAF_pruning == TRUE) { 
#       # Check if MAF is less than 0.05
#       if (p < min_MAF) {
#         # If MAF is less than 0.05, add the row index to the vector of rows to be removed
#         rows_to_remove <- c(rows_to_remove, j)
#       } else {
#         # Calculate expected heterozygosity using the provided function
#         expected_heterozygosity <- calculate_expected_heterozygosity(p)
#   
#         # Append the calculated expected heterozygosity to the list
#         window_H_e_list <- c(window_H_e_list, expected_heterozygosity)
#       } }
#     else {
#         # Calculate expected heterozygosity using the provided function
#         expected_heterozygosity <- calculate_expected_heterozygosity(p)
#   
#         # Append the calculated expected heterozygosity to the list
#         window_H_e_list <- c(window_H_e_list, expected_heterozygosity)
# 
#       
#     }
#     
#   }
# 
#   
#     if (length(rows_to_remove) > 0) {  # Check if rows_to_remove is not empty
#       # Remove rows from simulated_model_allele_freq_tables with MAF less than 0.05
#       table_data_filtered <- table_data[-rows_to_remove, ]
#       simulated_model_allele_freq_tables[[i]]$data <- table_data_filtered
#     }
# 
#     
#   
#   # Assign the calculated expected heterozygosity list to the H_e attribute of table_info
#   simulated_model_allele_freq_tables[[i]]$H_e_list <- window_H_e_list
#   simulated_model_allele_freq_tables[[i]]$Marker_based_Average_H_e <- mean(simulated_model_allele_freq_tables[[i]]$H_e_list)
#   
#   # Update the count of pruned entries for this simulation
#   pruned_counts[i] <- length(rows_to_remove)
# }
# 
# paste("Snps for each simulation, before the pruning:",datapoints_before_MAF_filtering)
# 
# paste("Pruned SNPs:",pruned_counts)
# 
# # View(simulated_model_allele_freq_tables)

```
#4: Computing window-based average expected heterozygosity Step 1:
Create genomic windows Step 2: Compute the average H_e for each genomic
window individually
## 4.1: Empirical data
### 4.1.2 Genomewide
```{r}
########################################################################
# Step 1: Create genomic windows
########################################################################
# Initialize an empty list to store windows for each empirical_genomewide
empirical_genomewide_windows <- list()

# Iterating through each empirical_genomewide
for (index in seq_along(empirical_genomewide_tables)) {
  # Get the SNP positions, chromosome positions, and H_e values for the current empirical_genomewide
  snp_positions <- empirical_genomewide_tables[[index]][["data"]][["POS1"]]
  chr_positions <- empirical_genomewide_tables[[index]][["data"]][["CHR"]]
  he_values <- empirical_genomewide_tables[[index]][["data"]][["H_e"]]
  # Determine unique chromosomes present in the data
  unique_chromosomes <- unique(chr_positions)
  
  # Iterate through each chromosome
  for (chromosome in unique_chromosomes) {
    # Filter SNP positions and H_e values for the current chromosome
    chr_snp_positions <- snp_positions[chr_positions == chromosome]
    chr_he_values <- he_values[chr_positions == chromosome]
    
    # Determine the number of windows needed for the current chromosome
    num_windows <- ceiling(max(chr_snp_positions) / window_size)
    # Initialize an empty list to store windows for the current chromosome
    chr_windows <- list()
    
    # Iterate through each window
    for (window_index in seq_len(num_windows)) {
      # Determine the start and end positions of the current window
      window_start <- (window_index - 1) * window_size + 1
      window_end <- window_start + window_size - 1
      # Find the indices of SNPs within the current window
      snp_indices <- which(chr_snp_positions >= window_start & chr_snp_positions <= window_end)
      
      # Skip the window if there are no SNPs within it
      if (length(snp_indices) == 0) {
        next
      }

      else {
          # Extract the SNP positions and H_e values for the current window
          window_snp_positions <- chr_snp_positions[snp_indices]
          window_he_values <- chr_he_values[snp_indices]
          # Create a sub-table for the current window
          window_data <- data.frame(POS = window_snp_positions, H_e = window_he_values)
          # Append the sub-table to the list of windows for the current chromosome
          chr_windows[[length(chr_windows) + 1]] <- window_data
    }
    }
    # Append the list of windows for the current chromosome to the main list
    empirical_genomewide_windows[[chromosome]] <- chr_windows
  }
}

# View the resulting windows
# View(empirical_genomewide_windows)
```


```{r}
#######################################################################
# Step 2: Compute the average H_e for each genomic window individually
#######################################################################
# Create a list to store the average H_e values for each empirical_genomewide
Empirical_genomewide_window_he_3_1 <- list()

# Iterate over each empirical_genomewide
for (i in seq_along(empirical_genomewide_windows)) {
    # Create a vector to store the average H_e values for each window in the current empirical_genomewide
    Empirical_genomewide_window_he_3_1[[i]] <- numeric(length(empirical_genomewide_windows[[i]]))
    
    # Iterate over each window in the current empirical_genomewide
    for (j in seq_along(empirical_genomewide_windows[[i]])) {
        # Calculate the average H_e value for the current window
        # Using "na.rm = TRUE" To remove windows not having SNP-markers in them
        window_avg_he <- mean(empirical_genomewide_windows[[i]][[j]][["H_e"]], na.rm = TRUE) 
        # Store the average H_e value for the current window
        Empirical_genomewide_window_he_3_1[[i]][j] <- window_avg_he
    }
    # Removing NaN values for the current empirical_genomewide
    # (Windows without any markers)
    Empirical_genomewide_window_he_3_1[[i]] <- Empirical_genomewide_window_he_3_1[[i]][!is.na(Empirical_genomewide_window_he_3_1[[i]])]

}
# View(Empirical_genomewide_window_he_3_1)
# Calculating the mean H_e value for each chromosome
# sapply applies the mean-function on each chr in Simulation_windows_avg_he_3_2
empirical_chromosomes_mean_he_window_based <- sapply(Empirical_genomewide_window_he_3_1, mean)

cat("Average H_e of each chromosome in the empirical dataset:") 
print(empirical_chromosomes_mean_he_window_based)

#################################################################################
# Step 3: Compute the genomwide average H_e for the empirical data (window-based)
#################################################################################
empirical_genomewides_mean_he_window_based <- mean(empirical_chromosomes_mean_he_window_based)
cat("Average H_e of the empirical dataset:")
print(empirical_genomewides_mean_he_window_based)
```
## 4.2: Simulated data (Neutral)
```{r}
########################################################################
# Step 1: Create genomic windows
########################################################################
# Initialize an empty list to store windows for each simulation
simulation_windows <- list()

# Iterating through each simulation
for (sim_index in seq_along(simulated_model_allele_freq_tables)) {
  # Get the SNP positions and H_e values for the current simulation
  snp_positions <- simulated_model_allele_freq_tables[[sim_index]][["data"]][["POS1"]]
  he_values <- simulated_model_allele_freq_tables[[sim_index]][["data"]][["H_e"]]
  # Determine the number of windows needed
  num_windows <- ceiling(max(snp_positions) / window_size)
  # Initialize an empty list to store windows for the current simulation
  sim_windows <- list()

  # Iterate through each window
  for (window_index in seq_len(num_windows)) {
    # Determine the start and end positions of the current window
    window_start <- (window_index - 1) * window_size +1
    window_end <- window_start + window_size - 1
    # Find the indices of SNPs within the current window
    snp_indices <- which(snp_positions >= window_start & snp_positions <= window_end)
    
    # Skip the window if there are no SNPs within it
    if (length(snp_indices) == 0) {
      next
    } else {
        # Extract the SNP positions and H_e values for the current window
        window_snp_positions <- snp_positions[snp_indices]
        window_he_values <- he_values[snp_indices]
        # Create a sub-table for the current window
        window_data <- data.frame(POS = window_snp_positions, H_e = window_he_values)
        # Append the sub-table to the list of windows for the current simulation
        sim_windows[[length(sim_windows) + 1]] <- window_data
    }
  }
  # Append the list of windows for the current simulation to the main list
  simulation_windows[[sim_index]] <- sim_windows
}
# View(simulation_windows)

#######################################################################
# Step 2: Compute the average H_e for each genomic window individually
#######################################################################
# Create a list to store the average H_e values for each simulation
Simulation_windows_avg_he_3_2 <- list()

# Iterate over each simulation
for (i in seq_along(simulation_windows)) {
    # Create a vector to store the average H_e values for each window in the current simulation
    Simulation_windows_avg_he_3_2[[i]] <- numeric(length(simulation_windows[[i]]))

    # Iterate over each window in the current simulation
    for (j in seq_along(simulation_windows[[i]])) {
        # Calculate the average H_e value for the current window
        # Using "na.rm = TRUE" To remove windows not having SNP-markers in them
        window_avg_he <- mean(simulation_windows[[i]][[j]][["H_e"]], na.rm = TRUE) 
        
        # Store the average H_e value for the current window
        Simulation_windows_avg_he_3_2[[i]][j] <- window_avg_he
    }
    # Removing NaN values for the current simulation
    # (Windows without any markers)
    Simulation_windows_avg_he_3_2[[i]] <- Simulation_windows_avg_he_3_2[[i]][!is.na(Simulation_windows_avg_he_3_2[[i]])]

}

# sim_id_to_follow <- 1
# 
# Windows_with_markers_after_MAF_pruning <- length(Simulation_windows_avg_he_3_2[[sim_id_to_follow]])
# cat("Number of 100kbp Windows with markers (after MAF-pruning), for simulation ",sim_id_to_follow, ": ",Windows_with_markers_after_MAF_pruning,"\n")
# 
# windows_with_he_01_or_less <- length(Simulation_windows_avg_he_3_2[[sim_id_to_follow]][Simulation_windows_avg_he_3_2[[sim_id_to_follow]] <= 0.1])
# cat("% of 100kbp Windows with H_e  <= 0.1:\n", round((windows_with_he_01_or_less/Windows_with_markers_after_MAF_pruning)*100 , 1), "\n")
# 
# fully_fixated_windows <- length(Simulation_windows_avg_he_3_2[[sim_id_to_follow]][Simulation_windows_avg_he_3_2[[sim_id_to_follow]] == 0])
# 
# cat("% Fully fixated windows, with average H_e = 0:\n", round((fully_fixated_windows/Windows_with_markers_after_MAF_pruning)*100 , 1), "\n")

# Calculate the mean H_e value for each simulation
# sapply applies the mean-function on each chr in Simulation_windows_avg_he_3_2 
simulations_mean_he_window_based <- sapply(Simulation_windows_avg_he_3_2, mean)
simulations_mean_he_window_based
```
# 5: Sweep test (Selection testing)
## 5.1 Computing 5th percentiles of H_e distribution
### 5.1.1 Empirical dataset - H_e
```{r}
# Check if pruning markers by MAF should be done
if (use_MAF_pruning == TRUE) {
    MAF_setting_text <- paste0("MAF >= ", min_MAF)
    # Remove rows where MAF < min_MAF
    hotspot_allele_freq_data <- hotspot_allele_freq_data[hotspot_allele_freq_data$MAF >= min_MAF, ]
} else {
    MAF_setting_text <- "No MAF-pruning"
}

# Determine the number of bins you want
num_bins <- 100  
histogram_title <-  paste0(empirical_breed," H_e Distribution (",MAF_setting_text,") - Windowsize: ",window_size,"bp ")
image_name <- paste0("H_e_distribution_",empirical_breed,".png")
# Save the plot as a PNG file with specified path
png(file.path(output_empirical_H_e_dir, image_name), width = 1200, height = 900, res = 120)
# Create a histogram with custom breaks
# Using unlist to 
hist(unlist(Empirical_genomewide_window_he_3_1), breaks = num_bins, main = histogram_title, xlab = "H_e")

# Calculate the 5th percentile
empirical_dataset_percentile_5 <- quantile(unlist(Empirical_genomewide_window_he_3_1), 0.05)
# Add a vertical line at the 5th percentile with increased thickness
abline(v = empirical_dataset_percentile_5, col = "blue", lty = 2, lwd = 2)
# Add a vertical line for the mean H_e with increased thickness
abline(v = empirical_genomewides_mean_he_window_based, col = "red", lty = 2, lwd = 2)
# Combine legends into one with matching line thickness
legend("topright", 
       legend = c(paste("5th percentile: ", round(empirical_dataset_percentile_5, H_e_decimals_plots)),
                  paste("Average H_e: ", round(empirical_genomewides_mean_he_window_based, H_e_decimals_plots))),
       col = c("blue", "red"), 
       lty = 2, 
       lwd = 2,
       inset = c(0, 0.02))
6
# Save the plot
dev.off()
# Initialize an empty data frame to store the results
empirical_5th_percentiles_df <- data.frame(Fifth_Percentile = empirical_dataset_percentile_5,Avg_H_e = empirical_genomewides_mean_he_window_based, stringsAsFactors = FALSE)
# Define the filename with the output directory path
filename <- file.path(output_empirical_H_e_dir, paste0("H_e_distribution_",empirical_breed,".tsv"))
# Saving the results in a .tsv-file
write.table(empirical_5th_percentiles_df, filename, sep = "\t", row.names = FALSE,quote = FALSE)
```
### 5.1.2: Simulated model - H_e distribution
*Using the first simulation from Simulation_windows_avg_he_3_2*
```{r}
setwd(output_simulated_model_H_e_dir)
# Determine the number of bins you want
num_bins <- 100  
# Initialize an empty data frame to store the results
sim_5th_percentiles_df <- data.frame(Simulation = character(), Fifth_Percentile = numeric(),Avg_H_e = numeric(), stringsAsFactors = FALSE)

# Loop through each simulation
for (i in seq_along(Simulation_windows_avg_he_3_2)) {
  # Create histogram title
  histogram_title <- paste0("Simulation ",i," H_e Distribution (",MAF_setting_text,") - Windowsize: ",window_size,"bp ")
  image_name <- paste0(sim_scenario_id,"_H_e_distribution_sim_",i,".png")
  # Save the plot as a PNG file with specified path
  png(file.path(output_simulated_model_H_e_dir, image_name), width = 1200, height = 900, res = 120)
  # Create a histogram with custom breaks
  hist(Simulation_windows_avg_he_3_2[[i]], breaks = num_bins, main = histogram_title, xlab = "H_e")
  # Calculate the 5th percentile
  percentile_5 <- quantile(Simulation_windows_avg_he_3_2[[i]], 0.05)
  simulated_model_allele_freq_tables[[i]]$H_e_5th_percentile <- percentile_5
  sim_window_based_H_e_avg <- mean(Simulation_windows_avg_he_3_2[[i]])
  # Add the simulation name and the fifth percentile value to the data frame
  sim_5th_percentiles_df[i, "Simulation"] <- paste("Simulation", i)
  sim_5th_percentiles_df[i, "Fifth_Percentile"] <- percentile_5
  sim_5th_percentiles_df[i, "Avg_H_e"] <- sim_window_based_H_e_avg
  # Add a vertical line at the 5th percentile
  abline(v = percentile_5, col = "blue", lty = 2)
  # Add a vertical line for the mean H_e with increased thickness
  abline(v = sim_window_based_H_e_avg, col = "red", lty = 2, lwd = 2)
  # Combine legends into one with matching line thickness
  legend("topright", 
       legend = c(paste("5th percentile: ", round(percentile_5, H_e_decimals_plots)),
                  paste("Average H_e: ", round(sim_window_based_H_e_avg, H_e_decimals_plots))),
       col = c("blue", "red"), 
       lty = 2, 
       lwd = 2,
       inset = c(0, 0.02))
  # Save the plot
  dev.off()
}
# Sort the data frame by Fifth_Percentile values in ascending order
sim_5th_percentiles_df <- sim_5th_percentiles_df[order(sim_5th_percentiles_df$Fifth_Percentile), ]

# Print the data frame
#print(sim_5th_percentiles_df)
kable(sim_5th_percentiles_df,row.names = FALSE)
# Define the filename with the output directory path
filename <- file.path(output_simulated_model_H_e_dir, paste0(sim_scenario_id,"_5th_percentiles_of_H_e_distribution.tsv"))
# Saving the results in a .tsv-file
write.table(sim_5th_percentiles_df, filename, sep = "\t", row.names = FALSE,quote = FALSE)
```
## SE Confidence Interval
```{r}
# Function to calculate standard error and its confidence interval
standard_error_confidence_interval_fun <- function(observed_data, confidence_level = 0.95) {
  # Calculate standard error
  n <- length(observed_data)
  standard_deviation <- sd(observed_data)
  standard_error <- standard_deviation / sqrt(n - 1)
  # Calculate confidence interval based on standard error
  alpha <- (1 - confidence_level) / 2
  margin_of_error <- qnorm(1 - alpha) * standard_error
  mean_estimate <- mean(observed_data)
  # Calculate the percentiles for the confidence interval
  confidence_interval_lower_bound <- mean_estimate - margin_of_error # 2.5th percentile (2σ)
  confidence_interval_upper_bound <- mean_estimate + margin_of_error # 97.5th percentile (2σ)
  # Return confidence interval
  return(c(confidence_interval_lower_bound, confidence_interval_upper_bound))
}
```
## 5.2: ROH-Hotspots Sweep test
Selection testing for the different ROH-Hotspot windows.
ROH-Hotspots with an average H_e lower than **or** equal to (\<=) the
5th percentile of the H_e-distribution for the simulated Simulated model,
**are ROH-Hotspots deemed as being under selection**.
```{r}
setwd(output_dir_sweep_test)
# Defining the dataframe that will store the selection testing results
ROH_hotspots_selection_testing <- data.frame(
  Name = character(),  
  Under_selection = character(),
  Window_based_Average_H_e = numeric() 
)
if ( nrow(sim_5th_percentiles_df) > 1 ){
  sim_H_e_5th_percentile_SE_CI <- standard_error_confidence_interval_fun(sim_5th_percentiles_df$Fifth_Percentile)
  sim_H_e_5th_lower_ci <- sim_H_e_5th_percentile_SE_CI[1]
  threshold_value <- sim_H_e_5th_lower_ci
} else {
  threshold_value <- sim_5th_percentiles_df$Fifth_Percentile
}

for (empirical_hotspot in names(empirical_hotspot_tables)) {
  # Check if Window_based_Average_H_e is <= percentile_5 for the current table (ROH-hotspot)
  if (empirical_hotspot_tables[[empirical_hotspot]][["Avg_He"]] <= threshold_value) {
    # # If under selection, append the name and Window_based_Average_H_e value with "Yes"
    ROH_hotspots_selection_testing <- rbind(ROH_hotspots_selection_testing, data.frame(
      Name = empirical_hotspot,
      Under_selection = "Yes",  # Marking as "Yes" if under selection
      Window_based_Average_H_e = empirical_hotspot_tables[[empirical_hotspot]][["Avg_He"]]
    ))
  } else {
    # If not under selection, append the name and Window_based_Average_H_e value with "No" 
    ROH_hotspots_selection_testing <- rbind(ROH_hotspots_selection_testing, data.frame(
      Name = empirical_hotspot,
      Under_selection = "No",  # Marking as "No" if not under selection
      Window_based_Average_H_e = empirical_hotspot_tables[[empirical_hotspot]][["Avg_He"]]
    ))
  }
}
# Sort the data frame by Window_based_Average_H_e values in ascending order
ROH_hotspots_selection_testing <- ROH_hotspots_selection_testing[order(ROH_hotspots_selection_testing$Window_based_Average_H_e), ]
# Create histogram title
histogram_title <- paste("ROH-Hotspots Distribution of Average H_e Values", i, "- Windowsize:", window_size, "bp")
image_name <- paste("ROH-Hotspots_Avg_H_e_distribution_",sim_scenario_id,"_comparison.png")
# Save the plot as a PNG file with specified path
png(file.path(output_dir_sweep_test, image_name), width = 1200, height = 900, res = 120)
# Plot histogram of hotspot Window_based_Average_H_e values
histogram <- hist(ROH_hotspots_selection_testing$Window_based_Average_H_e, breaks = 20, main = histogram_title, xlab = "H_e")
# Add a vertical line at the 5th percentile for simulated dataset
abline(v = min(sim_5th_percentiles_df$Fifth_Percentile), col = "red", lty = 2)
# Adding legend for the 5th percentile of the simulated dataset
legend("topright", legend = paste("Simulated (5th percentile) =", round(min(sim_5th_percentiles_df$Fifth_Percentile), 4)),
       col = "red", lty = 2, cex = 0.8, text.col = "black")

# Add a vertical line at the 5th percentile for empirical dataset
abline(v = empirical_dataset_percentile_5, col = "blue", lty = 2)
# Adding legend for the 5th percentile of the empirical dataset
legend("topleft", legend = paste("Empirical (5th percentile) =", round(empirical_dataset_percentile_5, 4)),
       col = "blue", lty = 2, cex = 0.8, text.col = "black")
# Close the PNG file
dev.off()
cat("ROH-hotspot selection testing results:\n")
kable(ROH_hotspots_selection_testing, row.names = FALSE)
# Formatting the threshold value for the output file, to 3 decimals
formatted_H_e_threshold_value <- round(threshold_value,5)
# Define the filename with the output directory path
filename <- file.path(output_dir_sweep_test, paste("ROH_hotspots_", sweep_test_type,"_",sim_scenario_id,"_H_E_thr_", formatted_H_e_threshold_value, ".tsv", sep = ""))
# Saving the results in a .tsv-file
write.table(ROH_hotspots_selection_testing, filename, sep = "\t", row.names = FALSE,quote = FALSE)
```