# config.sh – defines paths and parameters for the pipeline

# Specify the path to the Conda initialization script
conda_setup_script_path="" 
# conda_setup_script_path="/home/jonat/anaconda3/etc/profile.d/conda.sh" 

# Species of the empirical dataset (affects for instance species-specific options in PLINK)
export empirical_species="dog"

# Name of the empirical breed/species being analyzed
# It is assumed that the analyzed empirical dataset will be located in a folder with the value of this parameter (I.e: /data/raw/empirical/$empirical_breed or /data/raw/empirical/labrador_retriever for the example below)
export empirical_breed="labrador_retriever"

# Basename of the raw empirical dataset files
export empirical_raw_data_basename="LR_fs"

# Range of autosomal chromosomes to include in analysis. This value should be set according to the species being analyzed.  
# The format follows PLINK's chromosome specification (e.g., "1-38" for dog). 
export empirical_autosomal_chromosomes="1-38"

#�����������������������������
#� Mutation Rate �
#�����������������������������
# Set the estimated mutation rate for the studies species, to be used in the coalescent simulations and during the subsequent forward simulations if introduce_mutations=TRUE
# By default, the estimated mutation rate for dogs by (Lindblad-Toh et al.2005), 1 × 10−8 mutations per base pair per generation, is used.
export mutation_rate=1e-08

#�����������������������������
#� Chromosome-specific Recombination Rates �
#�����������������������������
# Define the chromosome specific recombination rates (in centiMorgan per Megabase) for the studies species.
# in the associative array; $chromosome_recombination_rates_cM_per_Mb below.

# If the studied empirical species is dogs, the predefined values below can be used.
# These comes from the canine linkage map produced by Wong et al. 2010.
# The source table for these values can be found here:
# https://www.researchgate.net/figure/Physical-and-genetic-lengths-of-dog-chromosomes_tbl3_40453037

declare -A chromosome_recombination_rates_cM_per_Mb=(
["chr1"]=0.68 ["chr2"]=0.88 ["chr3"]=0.69 ["chr4"]=0.75 ["chr5"]=0.85 
["chr6"]=0.71 ["chr7"]=0.75  ["chr8"]=0.83 ["chr9"]=1.11 ["chr10"]=0.78 
["chr11"]=0.84 ["chr12"]=0.87 ["chr13"]=0.90 ["chr14"]=0.89 ["chr15"]=0.80 
["chr16"]=0.90 ["chr17"]=0.89 ["chr18"]=0.88 ["chr19"]=0.93 ["chr20"]=0.93 
["chr21"]=1.02 ["chr22"]=0.84 ["chr23"]=1.02 ["chr24"]=1.13 ["chr25"]=1.05 
["chr26"]=1.22 ["chr27"]=1.18 ["chr28"]=1.31 ["chr29"]=1.15 ["chr30"]=1.36 
["chr31"]=1.29 ["chr32"]=1.37 ["chr33"]=1.53 ["chr34"]=1.17 ["chr35"]=2.15 
["chr36"]=1.54 ["chr37"]=1.38 ["chr38"]=2.22 	
)
#�����������������������������
#� Gene and Phenotype Mapping �
#�����������������������������
# Specify the path to a reference gene annotation file
export gene_annotations_filepath="$data_dir/preprocessed/empirical/gene_annotations/canFam3.ncbiRefSeq.gtf.gz" 

# Specify the path to the scraped (raw) OMIA phenotype file
# omia_scraped_phenotypes_data_filepath: This file is generated by the script: roh-island-simulation/code/OMIA_scraping/OMIA_phenotype_scraping.py
export omia_scraped_phenotypes_data_filepath="$data_dir/raw/empirical/omia_scraped_phene_data/OMIA_dog_phenotype_data_raw.csv"

# Specify the path to the OMIA phenotype .bed file
# omia_scraped_phenotypes_data_filepath: This file is generated by the script: roh-island-simulation/code/OMIA_scraping/create_phenotype_windows_bed.py
export omia_phenotypes_filepath="$data_dir/preprocessed/empirical/omia_phenotype_data/All_dog_phenotypes.bed"

# To list the Vertebrate Breed Ontology ID:s to use for associating phenotypes with the studied species, define it in
# the  'vertebrate_breed_ontology_ids' variable below. If you wish to use more than one VBO ID, the ID:s should be commaseparated as in
# the following example:  vertebrate_breed_ontology_ids="VBO_0200800,Unspecified".
# In the following example, VBO_0200800 refers to Labrador Retriever Dog, while "Unspecified" is an option to associate phenotypes
# with unknown (Unspecified) VBO ID to be able to be mapped to the studied species.
# A straight forward way to find the VBO ID of theg studied species, is by using the link below and replace "Labrador Retriever" with the studied species.'
#trieve the VBO ID for the studied breed/species, search on the following site: 
# https://ontobee.org/search?ontology=VBO&keywords=labrador+retriever&submit=Search+terms
export vertebrate_breed_ontology_ids="VBO_0200800,Unspecified" 




###################################################################################################
source $conda_setup_script_path  # Source Conda initialization script
# Activate the conda environment
conda activate roh_island_sim_env
# -------------[ Colors for nicer output ]-------------
RED='\033[0;31m'
YELLOW='\033[1;33m'
GREEN='\033[0;32m'
NC='\033[0m' # No Color
# -------------[ Main Logic Placeholder ]-------------
echo -e "\n${GREEN}[INFO]${NC} All required variables are set. Starting pipeline..."