% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Labrador Retriever 98 Gen Breed Formation - SNP chip: last\_breed\_formation\_generation - 50 N\_e bottleneck scenario},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Labrador Retriever 98 Gen Breed Formation - SNP chip:
last\_breed\_formation\_generation - 50 N\_e bottleneck scenario}
\author{}
\date{\vspace{-2.5em}2025-02-18}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{2}
\tableofcontents
}
\section{0: Preparation}\label{preparation}

Defining the input and output directories

\subsection{Loading libraries}\label{loading-libraries}

\begin{verbatim}
## Loading required package: knitr
\end{verbatim}

\begin{verbatim}
## Loading required package: ggplot2
\end{verbatim}

\begin{verbatim}
## Warning: package 'ggplot2' was built under R version 4.3.2
\end{verbatim}

\begin{verbatim}
## Loading required package: scatterplot3d
\end{verbatim}

\begin{verbatim}
## Warning: package 'scatterplot3d' was built under R version 4.3.1
\end{verbatim}

\begin{verbatim}
## Loading required package: RColorBrewer
\end{verbatim}

\section{Latex formatting function}\label{latex-formatting-function}

\section{Standard Error Confidence interval
function}\label{standard-error-confidence-interval-function}

\section{1: Causative variant (Selection
Models)}\label{causative-variant-selection-models}

\subsection{1.1: Variant fixation}\label{variant-fixation}

\subsubsection{1.1.3: Summary - Variant
fixation}\label{summary---variant-fixation}

Summary of (near) complete sweeps of the causative variant under
different selection models.

\begin{itemize}
\item
  Column 2 of the table shows the fixation probablity, i.e, the
  likelihood of a randomly selected causative variant reaching fixation
  at the predefined fixation-threshold (99\% by default) during a
  simulation run.
\item
  Column 3-5 display fixation time statistics from the conditional
  simulations regarding the average, minimum and maximum number of
  generations of required for fixation in theforward-in-time
  simulations.
\end{itemize}

The table is sorted in ascending order by selection coefficient.

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/Causative_variant_Fixation_summary.txt"
\end{verbatim}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 8\tabcolsep) * \real{0.2268}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 8\tabcolsep) * \real{0.2165}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 8\tabcolsep) * \real{0.1856}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 8\tabcolsep) * \real{0.1856}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 8\tabcolsep) * \real{0.1856}}@{}}
\toprule\noalign{}
\begin{minipage}[b]{\linewidth}\raggedright
Selection\_coefficient
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Fixation\_probability
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Avg\_Fixation\_time
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Min\_fixation\_time
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Max\_fixation\_time
\end{minipage} \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
s=0.075 & 2.3 & 80 & 56 & 97 \\
s=0.1 & 16.7 & 73 & 59 & 91 \\
s=0.125 & 45.9 & 71 & 48 & 92 \\
s=0.15 & 83.3 & 62 & 42 & 84 \\
s=0.2 & 98.0 & 48 & 38 & 70 \\
s=0.3 & 100.0 & 35 & 28 & 46 \\
s=0.4 & 100.0 & 27 & 21 & 36 \\
s=0.5 & 100.0 & 23 & 19 & 30 \\
s=0.6 & 100.0 & 20 & 15 & 22 \\
s=0.7 & 100.0 & 18 & 15 & 21 \\
s=0.8 & 100.0 & 16 & 12 & 19 \\
\end{longtable}

\subsection{1.2: Causative variant
windows}\label{causative-variant-windows}

\subsubsection{1.2.3: ROH frequency}\label{roh-frequency}

To see ROH-frequency plots of the created Causative Variant Windows,
uncomment the block of code between ``Start plot code'' and ``End plot
code'' and then rerun the script.

\subsection{1.3: Expected Heterozygosity - Causative Variant
Windows}\label{expected-heterozygosity---causative-variant-windows}

\subsection{1.4: Summary - Causative variant
windows}\label{summary---causative-variant-windows}

Summary of causative variant windows from the simulated selection
scenarios, presenting averages and confidence intervals (CI) for window
length, ROH frequency and expected heterozygosity. The table also
includes minimum and maximum ranges of ROH frequency values and shows
the average ROH frequency and its CI for the Original 100 kbp Causative
Variant Subwindow. The table is sorted in ascending order by selection
coefficient.

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/Causative_variant_windows_summary.txt"
\end{verbatim}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0424}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0424}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0678}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0678}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0381}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0763}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0763}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0381}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0381}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0975}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.1356}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.1356}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0339}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0551}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 28\tabcolsep) * \real{0.0551}}@{}}
\toprule\noalign{}
\begin{minipage}[b]{\linewidth}\raggedright
Sel.coeff
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Length\_Mb
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Length\_lower\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Length\_Upper\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
ROH\_freq
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
ROH\_freq\_lower\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
ROH\_freq\_upper\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Min\_freq
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Max\_freq
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
variant\_subwindow\_freq
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
variant\_subwindow\_freq\_lower\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
variant\_subwindow\_freq\_upper\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Avg\_H\_e
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
H\_e\_lower\_CI
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
H\_e\_upper\_CI
\end{minipage} \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
s=0.075 & 1.04 & 0.92 & 1.16 & 77.1 & 72.7 & 81.5 & 30.9 & 100 & 79.3 &
75.0 & 83.6 & 0.112 & 0.091 & 0.133 \\
s=0.1 & 1.11 & 0.96 & 1.26 & 77.8 & 73.1 & 82.4 & 27.6 & 100 & 79.2 &
74.7 & 83.7 & 0.118 & 0.094 & 0.141 \\
s=0.125 & 1.11 & 0.96 & 1.26 & 82.2 & 77.2 & 87.3 & 24.8 & 100 & 83.9 &
78.9 & 88.8 & 0.100 & 0.076 & 0.124 \\
s=0.15 & 1.26 & 1.10 & 1.42 & 87.0 & 83.4 & 90.5 & 50.6 & 100 & 88.6 &
85.1 & 92.1 & 0.092 & 0.068 & 0.115 \\
s=0.2 & 1.41 & 1.17 & 1.64 & 87.3 & 83.8 & 90.7 & 49.1 & 100 & 88.9 &
85.6 & 92.3 & 0.093 & 0.072 & 0.114 \\
s=0.3 & 1.58 & 1.34 & 1.82 & 93.4 & 91.3 & 95.5 & 68.8 & 100 & 94.6 &
92.6 & 96.7 & 0.086 & 0.066 & 0.105 \\
s=0.4 & 1.96 & 1.67 & 2.25 & 93.5 & 90.7 & 96.2 & 34.2 & 100 & 94.8 &
92.1 & 97.5 & 0.069 & 0.053 & 0.084 \\
s=0.5 & 2.22 & 1.90 & 2.54 & 97.2 & 96.1 & 98.2 & 80.6 & 100 & 98.4 &
97.4 & 99.5 & 0.084 & 0.058 & 0.110 \\
s=0.6 & 2.06 & 1.79 & 2.33 & 96.2 & 94.4 & 98.0 & 67.9 & 100 & 97.5 &
95.8 & 99.2 & 0.063 & 0.046 & 0.081 \\
s=0.7 & 2.39 & 2.05 & 2.74 & 97.7 & 96.7 & 98.6 & 75.5 & 100 & 98.8 &
97.9 & 99.7 & 0.058 & 0.038 & 0.078 \\
s=0.8 & 2.58 & 2.19 & 2.96 & 96.1 & 93.3 & 98.9 & 29.4 & 100 & 97.1 &
94.3 & 99.8 & 0.084 & 0.059 & 0.109 \\
\end{longtable}

\section{2: ROH-Frequency}\label{roh-frequency-1}

\subsection{2.1: Autosomal genomewide
ROH-frequencies}\label{autosomal-genomewide-roh-frequencies}

\subsection{2.2: ROH hotspot threshold
Summary}\label{roh-hotspot-threshold-summary}

The following table shows a comparison between the empirical ROH
hotspot-threshold and each simulation model type. Lower CI and Upper CI
correspond to the standard-error based confidence intervals of the ROH
hotspot threshold with a confidence level of 95\%.

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/ROH-hotspot_threshold_comparison.txt"
\end{verbatim}

\begin{longtable}[]{@{}lrrr@{}}
\toprule\noalign{}
Model & Avg\_ROH\_hotspot\_threshold & Lower\_CI & Upper\_CI \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Neutral & 54.0 & 50.7 & 57.4 \\
Empirical & 55.4 & NA & NA \\
s=0.075 & 71.0 & 68.0 & 74.1 \\
s=0.1 & 72.6 & 68.5 & 76.7 \\
s=0.125 & 75.1 & 70.8 & 79.3 \\
s=0.15 & 79.4 & 75.8 & 83.0 \\
s=0.2 & 83.1 & 79.4 & 86.8 \\
s=0.3 & 89.3 & 86.4 & 92.2 \\
s=0.4 & 92.6 & 90.1 & 95.1 \\
s=0.6 & 95.6 & 93.5 & 97.7 \\
s=0.5 & 96.5 & 94.8 & 98.2 \\
s=0.7 & 96.9 & 95.3 & 98.5 \\
s=0.8 & 98.0 & 96.7 & 99.3 \\
\end{longtable}

\section{3: Inbreeding coefficient}\label{inbreeding-coefficient}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.1: Empirical data-1.pdf}

\begin{verbatim}
## Overall Average F_ROH for  labrador_retriever : 0.2333818 //n
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.2: Neutral Model-1.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations: 0.2331937 //n
\end{verbatim}

\begin{verbatim}
## [1] "Standard Error 95% Confidence Interval: [0.22597765002533, 0.240409754823155]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-1.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s0075_chr1 : 0.2457452 //n[1] "standard error-based 95% Confidence Interval: [0.239008545960154, 0.252481917070149]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-2.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s01_chr1 : 0.247888 //n[1] "standard error-based 95% Confidence Interval: [0.242112669535767, 0.253663418949082]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-3.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s0125_chr1 : 0.2505926 //n[1] "standard error-based 95% Confidence Interval: [0.243284359917288, 0.257900848567561]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-4.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s015_chr1 : 0.2485584 //n[1] "standard error-based 95% Confidence Interval: [0.242527097994766, 0.254589697156749]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-5.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s02_chr1 : 0.2491423 //n[1] "standard error-based 95% Confidence Interval: [0.243150463048794, 0.255134104223933]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-6.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s03_chr1 : 0.2607497 //n[1] "standard error-based 95% Confidence Interval: [0.253973622180261, 0.267525811759133]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-7.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s04_chr1 : 0.2664084 //n[1] "standard error-based 95% Confidence Interval: [0.260592203730001, 0.27222464717909]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-8.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s05_chr1 : 0.2841088 //n[1] "standard error-based 95% Confidence Interval: [0.276958107715148, 0.291259418345458]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-9.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s06_chr1 : 0.2892974 //n[1] "standard error-based 95% Confidence Interval: [0.281786042544946, 0.296808739273235]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-10.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s07_chr1 : 0.3002323 //n[1] "standard error-based 95% Confidence Interval: [0.293042526021229, 0.307422059433317]"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/3.3: Selection Models-11.pdf}

\begin{verbatim}
## Point estimate F_ROH across all 20 simulations for  selection_model_s08_chr1 : 0.2925508 //n[1] "standard error-based 95% Confidence Interval: [0.285566916937773, 0.299534708516772]"
\end{verbatim}

\subsection{3.4: F\_ROH summary}\label{f_roh-summary}

This section shows a comparison of population average inbreeding
coefficient (F\_ROH) between each simulation type and the empirical
data. The displayed inbreeding coefficients for the different simulation
types is the point estimate FROH across the technical replicates. Lower
CI and Upper CI correspond to the sample standard-error based confidence
intervals of these population F\_ROH, with a confidence level of 95\%.

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/F_ROH_comparison.txt"
\end{verbatim}

\begin{longtable}[]{@{}lrrr@{}}
\toprule\noalign{}
Model & F\_ROH & Lower\_CI & Upper\_CI \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Neutral & 0.233 & 0.226 & 0.240 \\
Empirical & 0.233 & NA & NA \\
s=0.075 & 0.246 & 0.239 & 0.252 \\
s=0.1 & 0.248 & 0.242 & 0.254 \\
s=0.15 & 0.249 & 0.243 & 0.255 \\
s=0.2 & 0.249 & 0.243 & 0.255 \\
s=0.125 & 0.251 & 0.243 & 0.258 \\
s=0.3 & 0.261 & 0.254 & 0.268 \\
s=0.4 & 0.266 & 0.261 & 0.272 \\
s=0.5 & 0.284 & 0.277 & 0.291 \\
s=0.6 & 0.289 & 0.282 & 0.297 \\
s=0.8 & 0.293 & 0.286 & 0.300 \\
s=0.7 & 0.300 & 0.293 & 0.307 \\
\end{longtable}

\section{4: Expected Heterozygosity}\label{expected-heterozygosity}

\subsection{4.4: Genomewide Average H\_e
Summary}\label{genomewide-average-h_e-summary}

Summary of the mean-point estimation of the genome-wide average expected
heterozygosity (Hₑ), calculated across all 100 kbp binned genomic
regions, with standard error-based confidence intervals.

\begin{longtable}[]{@{}lrrr@{}}
\toprule\noalign{}
Model & H\_e & Lower\_CI & Upper\_CI \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Empirical & 0.316 & NA & NA \\
s=0.8 & 0.320 & 0.318 & 0.322 \\
s=0.7 & 0.321 & 0.318 & 0.323 \\
s=0.5 & 0.322 & 0.319 & 0.324 \\
s=0.4 & 0.324 & 0.322 & 0.325 \\
s=0.6 & 0.324 & 0.321 & 0.327 \\
s=0.2 & 0.326 & 0.324 & 0.328 \\
s=0.3 & 0.326 & 0.323 & 0.328 \\
s=0.15 & 0.327 & 0.325 & 0.329 \\
s=0.1 & 0.328 & 0.326 & 0.330 \\
s=0.075 & 0.329 & 0.327 & 0.331 \\
s=0.125 & 0.329 & 0.327 & 0.331 \\
Neutral & 0.330 & 0.329 & 0.332 \\
\end{longtable}

\subsection{4.5: Genomewide 5th percentile of Expected Heterozygosity
Summary}\label{genomewide-5th-percentile-of-expected-heterozygosity-summary}

Summary of the mean point estimation of expected heterozygosity (He) 5th
percentile values with standard error-based confidence intervals.

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/Expected_Heterozygosity_Summary.txt"
\end{verbatim}

\begin{longtable}[]{@{}lrrr@{}}
\toprule\noalign{}
Model & H\_e\_5th\_percentile & Lower\_CI & Upper\_CI \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
s=0.7 & 0.136 & 0.130 & 0.141 \\
s=0.5 & 0.138 & 0.132 & 0.143 \\
s=0.8 & 0.139 & 0.134 & 0.144 \\
s=0.6 & 0.140 & 0.135 & 0.145 \\
s=0.4 & 0.148 & 0.144 & 0.152 \\
Empirical & 0.151 & NA & NA \\
s=0.3 & 0.153 & 0.149 & 0.157 \\
s=0.2 & 0.157 & 0.154 & 0.160 \\
s=0.125 & 0.158 & 0.154 & 0.161 \\
s=0.15 & 0.158 & 0.154 & 0.162 \\
s=0.075 & 0.160 & 0.156 & 0.164 \\
s=0.1 & 0.160 & 0.156 & 0.164 \\
Neutral & 0.167 & 0.164 & 0.171 \\
\end{longtable}

\section{5. OMIA Phenotypes}\label{omia-phenotypes}

\subsubsection{5.1 All Breed-Specific Non-Defect (Non-Disease-Related)
Phenotypes}\label{all-breed-specific-non-defect-non-disease-related-phenotypes}

\begin{table}
\caption{\label{tab:5.1 All Breed-Specific Non-Defect (Non-Disease-Related) Phenotypes}Table showing all Non-Defect (Non-Disease-Related) breed-related phenotypes, discovered on OMIA}

\begin{longtable}{r|r|r|l|l|l|l|l|l|l|l|l}
\hline
CHR & POS1 & POS2 & PHENE & PHENE\_CATEGORY & SINGLE\_GENE\_TRAIT\_OR\_DISORDER & DISEASE\_RELATED & GENE\_SYMBOL & GENE\_DESCRIPTION & PHENE\_URL & GENE\_DETAILS\_URL & BREEDS\\
\hline
1 & 24326030 & 24331495 & Reduced hair shedding, MC5R-related & Integument (skin) phene & yes & no & MC5R & melanocortin 5 receptor & https://www.omia.org/OMIA002750/9615/ & https://www.omia.org/gene388250420/ & Unspecified\\
\hline
1 & 111055483 & 111058288 & Hyperlipidaemia/atherosclerosis, APOE-related & Homeostasis / metabolism phene & yes & no & APOE & apolipoprotein E & https://www.omia.org/OMIA002063/9615/ & https://www.omia.org/gene388244913/ & Unspecified\\
\hline
3 & 92125265 & 92275341 & Height, LCORL-associated body-size variation & Growth / size / body region phene & no & no & LCORL & ligand dependent nuclear receptor corepressor-like & https://www.omia.org/OMIA002246/9615/ & https://www.omia.org/gene388307121/ & Unspecified\\
\hline
5 & 63922271 & 63923224 & Coat colour, extension & Pigmentation phene & yes & no & MC1R & melanocortin 1 receptor (alpha melanocyte stimulating hormone receptor) & https://www.omia.org/OMIA001199/9615/ & https://www.omia.org/gene489652/ & Unspecified\\
\hline
5 & 63922271 & 63923224 & Coat colour, grizzle & Pigmentation phene & yes & no & MC1R & melanocortin 1 receptor (alpha melanocyte stimulating hormone receptor) & https://www.omia.org/OMIA001495/9615/ & https://www.omia.org/gene489652/ & Unspecified\\
\hline
5 & 63922271 & 63923224 & Coat colour, melanistic mask & Pigmentation phene & yes & no & MC1R & melanocortin 1 receptor (alpha melanocyte stimulating hormone receptor) & https://www.omia.org/OMIA001590/9615/ & https://www.omia.org/gene489652/ & Unspecified\\
\hline
\end{longtable}
\end{table}

\subsection{5.2 Species-Specific Phenotypes Overlapping With The
Detected Empirical ROH
Hotspot(s)}\label{species-specific-phenotypes-overlapping-with-the-detected-empirical-roh-hotspots}

\begin{verbatim}
## No non-defect (non-disease) phenotypes were found to overlap with the empirical ROH hotspot regions.
\end{verbatim}

\section{6: Summary}\label{summary}

\subsection{6.0: Detected Empirical ROH
hotspots}\label{detected-empirical-roh-hotspots}

Summary of ROH hotspot windows found in the empirical dataset.

\begin{longtable}[]{@{}lrrrr@{}}
\caption{Detected ROH hotspots from the empirical
dataset}\tabularnewline
\toprule\noalign{}
Name & Chr & Start & End & Length\_Mb \\
\midrule\noalign{}
\endfirsthead
\toprule\noalign{}
Name & Chr & Start & End & Length\_Mb \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Hotspot\_chr1\_window\_1 & 1 & 21900001 & 22600000 & 0.7 \\
Hotspot\_chr6\_window\_1 & 6 & 23600001 & 25000000 & 1.4 \\
Hotspot\_chr8\_window\_1 & 8 & 1500001 & 2700000 & 1.2 \\
Hotspot\_chr11\_window\_1 & 11 & 35400001 & 39400000 & 4.0 \\
Hotspot\_chr11\_window\_2 & 11 & 40200001 & 57200000 & 17.0 \\
Hotspot\_chr13\_window\_1 & 13 & 37300001 & 38500000 & 1.2 \\
Hotspot\_chr15\_window\_1 & 15 & 17000001 & 18700000 & 1.7 \\
Hotspot\_chr24\_window\_1 & 24 & 24600001 & 25900000 & 1.3 \\
Hotspot\_chr28\_window\_1 & 28 & 24100001 & 25500000 & 1.4 \\
\end{longtable}

\subsection{6.1: General comparison}\label{general-comparison}

\subsubsection{6.1.1: ROH-hotspot threshold
comparison}\label{roh-hotspot-threshold-comparison}

\begin{verbatim}
## 
##  ROH-hotspot threshold comparison between the different datasets
\end{verbatim}

\begin{longtable}[]{@{}lrrr@{}}
\toprule\noalign{}
Model & Avg\_ROH\_hotspot\_threshold & Lower\_CI & Upper\_CI \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Neutral & 54.0 & 50.7 & 57.4 \\
Empirical & 55.4 & NA & NA \\
s=0.075 & 71.0 & 68.0 & 74.1 \\
s=0.1 & 72.6 & 68.5 & 76.7 \\
s=0.125 & 75.1 & 70.8 & 79.3 \\
s=0.15 & 79.4 & 75.8 & 83.0 \\
s=0.2 & 83.1 & 79.4 & 86.8 \\
s=0.3 & 89.3 & 86.4 & 92.2 \\
s=0.4 & 92.6 & 90.1 & 95.1 \\
s=0.6 & 95.6 & 93.5 & 97.7 \\
s=0.5 & 96.5 & 94.8 & 98.2 \\
s=0.7 & 96.9 & 95.3 & 98.5 \\
s=0.8 & 98.0 & 96.7 & 99.3 \\
\end{longtable}

\subsubsection{6.1.2: F\_ROH comparison}\label{f_roh-comparison}

This section shows a plot comparing the population F\_ROH (y-axis)
between each simulation type (x-axis) and the empirical data.

The dashed horizontal red line represents the population FROH for
Labrador Retriever. The circle for each simulation type represents the
point estimate FROH across all simulations. The upper and lower error
margins of these point estimates represents the lower and upper standard
error-based confidence intervals with a confidence level of 95\%

\begin{verbatim}
## pdf 
##   2
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.1.2: F_ROH comparison-1.pdf}

\begin{verbatim}
## Models where empirical F_ROH is within CI:
\end{verbatim}

\begin{verbatim}
## [1] "Neutral"
\end{verbatim}

\begin{verbatim}
## 
## Models where empirical F_ROH is outside CI:
\end{verbatim}

\begin{verbatim}
##  [1] "s=0.075" "s=0.1"   "s=0.125" "s=0.15"  "s=0.2"   "s=0.3"   "s=0.4"  
##  [8] "s=0.5"   "s=0.6"   "s=0.7"   "s=0.8"
\end{verbatim}

\begin{longtable}[]{@{}lrrr@{}}
\toprule\noalign{}
Model & F\_ROH & Lower\_CI & Upper\_CI \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
Neutral & 0.233 & 0.226 & 0.240 \\
Empirical & 0.233 & NA & NA \\
s=0.075 & 0.246 & 0.239 & 0.252 \\
s=0.1 & 0.248 & 0.242 & 0.254 \\
s=0.15 & 0.249 & 0.243 & 0.255 \\
s=0.2 & 0.249 & 0.243 & 0.255 \\
s=0.125 & 0.251 & 0.243 & 0.258 \\
s=0.3 & 0.261 & 0.254 & 0.268 \\
s=0.4 & 0.266 & 0.261 & 0.272 \\
s=0.5 & 0.284 & 0.277 & 0.291 \\
s=0.6 & 0.289 & 0.282 & 0.297 \\
s=0.8 & 0.293 & 0.286 & 0.300 \\
s=0.7 & 0.300 & 0.293 & 0.307 \\
\end{longtable}

\subsection{6.2: Selection Testing of Empirical
ROH-Hotspots}\label{selection-testing-of-empirical-roh-hotspots}

\subsubsection{6.2.1: Selection test (Sweep
test)}\label{selection-test-sweep-test}

This section shows the selection testing results of the different ROH
hotspots based on the window average (He). The Table is sorted by
expected heterozygosity in ascending order.

The threshold value for selection used in this selection test, comes
from the lower boundary of the standard error-based confidence interval
for the 5th percentile extreme value of the neutral model simulations.

\begin{verbatim}
## [1] "Selection test results"
\end{verbatim}

\begin{verbatim}
## [1] "ROH-hotspot windows with an mean H_e Value lower or equal to the lower confidence interval of the fifth percentile of the neutral model are classified as candidate regions for selection"
\end{verbatim}

\begin{verbatim}
## [1] "5th percentile of the neutral model is: 0.16359"
\end{verbatim}

\begin{longtable}[]{@{}llr@{}}
\toprule\noalign{}
Name & Under\_selection & Window\_based\_Average\_H\_e \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
hotspot\_chr28\_window\_1 & Yes & 0.118 \\
hotspot\_chr1\_window\_1 & Yes & 0.144 \\
hotspot\_chr24\_window\_1 & No & 0.186 \\
hotspot\_chr13\_window\_1 & No & 0.190 \\
hotspot\_chr11\_window\_2 & No & 0.205 \\
hotspot\_chr15\_window\_1 & No & 0.223 \\
hotspot\_chr6\_window\_1 & No & 0.232 \\
hotspot\_chr11\_window\_1 & No & 0.233 \\
hotspot\_chr8\_window\_1 & No & 0.307 \\
\end{longtable}

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/ROH_hotspots_Selection_testing_neutral_model_H_E_threshold_0.164.csv.txt"
\end{verbatim}

\begin{verbatim}
## [1] "ROH hotspots marked as candidate regions for selection:"
\end{verbatim}

\begin{longtable}[]{@{}llr@{}}
\toprule\noalign{}
Name & Under\_selection & Window\_based\_Average\_H\_e \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
hotspot\_chr28\_window\_1 & Yes & 0.118 \\
hotspot\_chr1\_window\_1 & Yes & 0.144 \\
\end{longtable}

\subsubsection{(Optional) 6.2.2: Sweep Test of Causative Variant
Windows}\label{optional-6.2.2-sweep-test-of-causative-variant-windows}

This Optional quality control step verifies whether the Causative
Variant Windows in the simulated selection models are accurately
identified as candidate regions for selection based on the results of
the sweep test.

\begin{verbatim}
## [1] "C:/Users/jonat/GitHub/roh-island-simulation/results_chr1_3185_Ne_50TC/Pipeline_results/Causative_windows_under_selection.txt"
\end{verbatim}

\begin{longtable}[]{@{}lrrrl@{}}
\toprule\noalign{}
Model & H\_e & Lower\_CI & Upper\_CI & Under\_Selection \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
s=0.7 & 0.058 & 0.038 & 0.078 & Yes \\
s=0.6 & 0.063 & 0.046 & 0.081 & Yes \\
s=0.4 & 0.069 & 0.053 & 0.084 & Yes \\
s=0.5 & 0.084 & 0.058 & 0.110 & Yes \\
s=0.8 & 0.084 & 0.059 & 0.109 & Yes \\
s=0.3 & 0.086 & 0.066 & 0.105 & Yes \\
s=0.15 & 0.092 & 0.068 & 0.115 & Yes \\
s=0.2 & 0.093 & 0.072 & 0.114 & Yes \\
s=0.125 & 0.100 & 0.076 & 0.124 & Yes \\
s=0.075 & 0.112 & 0.091 & 0.133 & Yes \\
s=0.1 & 0.118 & 0.094 & 0.141 & Yes \\
Neutral & 0.167 & 0.164 & 0.171 & No \\
\end{longtable}

\subsection{6.3 Functional Assessment of Candidate Regions for
Selection}\label{functional-assessment-of-candidate-regions-for-selection}

\subsubsection{6.3.1 Genes Identified in Candidate
Regions}\label{genes-identified-in-candidate-regions}

\begin{verbatim}
## The following gene annotation file was used to map genes to the empirical candidate region(s) for selection: roh-island-simulation/data/preprocessed/empirical/gene_annotations/canFam3.ncbiRefSeq.gtf.gz
\end{verbatim}

\begin{verbatim}
## The following table(s) display genes overlapping with the ROH hotspot candidate region(s) for selection, in the following order:
##  hotspot_chr1_window_1 hotspot_chr28_window_1
\end{verbatim}

\begin{table}
\caption{\label{tab:6.3.1 Genes Identified in Candidate Regions}Table showing all genes overlapping with the candidate region(s) for selection}

\begin{tabular}{l|l|l|l|l}
\hline
CHR & POS1 & POS2 & Gene & Overlap\_Percentage\\
\hline
1 & 21788833 & 22494849 & DCC & 84.2542\\
\hline
\end{tabular}
\begin{tabular}{l|l|l|l|l}
\hline
CHR & POS1 & POS2 & Gene & Overlap\_Percentage\\
\hline
28 & 23966079 & 24162775 & TCF7L2 & 31.91457\\
\hline
28 & 24167153 & 24178702 & LOC111093006 & 100.00000\\
\hline
28 & 24228792 & 24234735 & LOC111093005 & 100.00000\\
\hline
28 & 24235208 & 24244092 & LOC106557942 & 100.00000\\
\hline
28 & 24245577 & 24246632 & LOC111093004 & 100.00000\\
\hline
28 & 24427734 & 24431660 & LOC111093054 & 100.00000\\
\hline
28 & 24489665 & 24522336 & HABP2 & 100.00000\\
\hline
28 & 24522285 & 24592267 & NRAP & 100.00000\\
\hline
28 & 24599866 & 24630018 & CASP7 & 100.00000\\
\hline
28 & 24643803 & 24685947 & PLEKHS1 & 100.00000\\
\hline
28 & 24704044 & 24724561 & DCLRE1A & 100.00000\\
\hline
28 & 24724697 & 24782600 & NHLRC2 & 100.00000\\
\hline
28 & 24749570 & 24907478 & LOC106557950 & 100.00000\\
\hline
28 & 24905369 & 24906950 & LOC111092962 & 100.00000\\
\hline
28 & 24908224 & 24909684 & ADRB1 & 100.00000\\
\hline
28 & 24963035 & 25022996 & CCDC186 & 100.00000\\
\hline
28 & 25026306 & 25028126 & LOC111093085 & 100.00000\\
\hline
28 & 25028195 & 25081300 & TDRD1 & 100.00000\\
\hline
28 & 25089499 & 25136879 & VWA2 & 100.00000\\
\hline
28 & 25137235 & 25228454 & AFAP1L2 & 100.00000\\
\hline
28 & 25178432 & 25184221 & LOC111093074 & 100.00000\\
\hline
28 & 25260323 & 25543369 & ABLIM1 & 84.67781\\
\hline
\end{tabular}
\end{table}

\subsubsection{6.3.2 Non-Defect (Non-Disease-Related) Phenotypes
Identified in Candidate
Regions}\label{non-defect-non-disease-related-phenotypes-identified-in-candidate-regions}

The following section displays the results from the search for known
favorable/non-defect (non-disease-related) phenotypes from OMIA,
overlapping with ROH hotspot candidate regions for selection.

This search is breed-specific and the Vertebrade Breed Ontology (VBO)
ID:s used, can be defined in the `vertebrate\_breed\_ontology\_ids'
parameter of the main pipeline script.

\begin{verbatim}
##  The following Vertebrate Breed Ontology (VBO) ID(s) were used to map phenotypes from OMIA with each candidate region for selection: VBO_0200800,Unspecified
\end{verbatim}

\begin{verbatim}
## The following OMIA phenotype file was used to map phenotypes to the empirical candidate region(s) for selection: C:/Users/jonat/GitHub/roh-island-simulation/data/preprocessed/empirical/omia_phenotype_data/All_dog_phenotypes.bed
\end{verbatim}

\begin{verbatim}
## Result: No Non-Defect (Non-Disease-Related) phenotypes was discovered for the ROH hotspot candidate region(s) for selection
\end{verbatim}

\subsection{6.4: Selection Strength Estimation of Empirical ROH Hotspot
Candidate Region(s) for
selection}\label{selection-strength-estimation-of-empirical-roh-hotspot-candidate-regions-for-selection}

In this section, the selection coefficients of empirical ROH-Hotspots,
identified as candidate regions for selection, are estimated by
comparing these regions with the Causative Variant Windows of the
simulated selection models. By identifying the selection model that
produces Causative Variant Windows most similar to the empirical region
in terms of: * ROH frequency * Expected heterozygosity * Window length
One can infer that the selection coefficient of the empirical candidate
region likely corresponds to the selection coefficient of the matched
selection model.

\subsubsection{6.4.1: Comparison of ROH-Hotspots and Causative Variant
Windows}\label{comparison-of-roh-hotspots-and-causative-variant-windows}

Comparison of the ROH hotspot candidate regions for selection in the
empirical dataset with the causative variant windows from the different
selection models simulating a hard sweep scenario. The table is sorted
by the selection strength of the models in descending alphanumerical
order.

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.1608}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.0769}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.1119}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.1119}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.0629}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.1259}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.1259}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.0420}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.0909}}
  >{\raggedleft\arraybackslash}p{(\columnwidth - 18\tabcolsep) * \real{0.0909}}@{}}
\toprule\noalign{}
\begin{minipage}[b]{\linewidth}\raggedright
Model
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Lengths\_Mb
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Length\_lower\_ci
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
Length\_upper\_ci
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
ROH\_Freq
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
ROH\_freq\_lower\_ci
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
ROH\_freq\_upper\_ci
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
H\_e
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
H\_e\_lower\_ci
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedleft
H\_e\_upper\_ci
\end{minipage} \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
s=0.8 & 2.58 & 2.19 & 2.96 & 96.1 & 93.3 & 98.9 & 0.084 & 0.059 &
0.109 \\
s=0.7 & 2.39 & 2.05 & 2.74 & 97.7 & 96.7 & 98.6 & 0.058 & 0.038 &
0.078 \\
s=0.6 & 2.06 & 1.79 & 2.33 & 96.2 & 94.4 & 98.0 & 0.063 & 0.046 &
0.081 \\
s=0.5 & 2.22 & 1.90 & 2.54 & 97.2 & 96.1 & 98.2 & 0.084 & 0.058 &
0.110 \\
s=0.4 & 1.96 & 1.67 & 2.25 & 93.5 & 90.7 & 96.2 & 0.069 & 0.053 &
0.084 \\
s=0.3 & 1.58 & 1.34 & 1.82 & 93.4 & 91.3 & 95.5 & 0.086 & 0.066 &
0.105 \\
s=0.2 & 1.41 & 1.17 & 1.64 & 87.3 & 83.8 & 90.7 & 0.093 & 0.072 &
0.114 \\
s=0.15 & 1.26 & 1.10 & 1.42 & 87.0 & 83.4 & 90.5 & 0.092 & 0.068 &
0.115 \\
s=0.125 & 1.11 & 0.96 & 1.26 & 82.2 & 77.2 & 87.3 & 0.100 & 0.076 &
0.124 \\
s=0.1 & 1.11 & 0.96 & 1.26 & 77.8 & 73.1 & 82.4 & 0.118 & 0.094 &
0.141 \\
s=0.075 & 1.04 & 0.92 & 1.16 & 77.1 & 72.7 & 81.5 & 0.112 & 0.091 &
0.133 \\
Hotspot\_chr28\_window\_1 & 1.40 & NA & NA & 75.0 & NA & NA & 0.118 & NA
& NA \\
Hotspot\_chr1\_window\_1 & 0.70 & NA & NA & 56.2 & NA & NA & 0.144 & NA
& NA \\
\end{longtable}

\subsubsection{6.4.2: 3D Plot Comparison of Mean
Values}\label{d-plot-comparison-of-mean-values}

3D-scatterplot comparing the ROH hotspot selection-candidate regions of
the empirical dataset with the causative variant windows from the
different selection models (monochronic color).

The x-axis shows the average ROH frequency of 100 kbp windows in the
region (percentage); the y-axis represents the average He of 100 kbp
windows in the region; and the z-axis represents the average length of
the respective regions (megabases). Vertical lines have been added for
each data point, projecting the regions onto the x-y plane to enhance
the clarity of the plot.
\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.2: 3D Plot Comparison of Mean Values-1.pdf}

\subsubsection{6.4.3: Visualization of Confidence Intervals and
Means}\label{visualization-of-confidence-intervals-and-means}

In this section, plots of the standard error-based confidence intervals,
alongside mean values, are created to account for the distribution of
values, rather than relying solely on the mean (as is the case in the
3D-plot). These plots are generated individually for each empirical
candidate region for selection, allowing a more detailed comparison
between each empirical ROH-Hotspot and the Causative Variant Windows of
the simulated selection models.

\paragraph{6.4.3.1: Standard Error CI for
H\_e}\label{standard-error-ci-for-h_e}

The following section displays plot(s) comparing the average expected
heterozygosity of the empirical ROH hotspot(s) with the corresponding
standard error-based confidence intervals for the different causative
variant windows.

\begin{verbatim}
## Models where empirical H_e is within CI for Hotspot_chr28_window_1 :
## [1] "s=0.075" "s=0.1"   "s=0.125"
## 
## Models where empirical H_e is outside CI for Hotspot_chr28_window_1 :
## [1] "s=0.15" "s=0.2"  "s=0.3"  "s=0.4"  "s=0.5"  "s=0.6"  "s=0.7"  "s=0.8"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.3.1: Standard Error CI for H_e-1.pdf}
\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.3.1: Standard Error CI for H_e-2.pdf}

\begin{verbatim}
## Models where empirical H_e is within CI for Hotspot_chr1_window_1 :
## character(0)
## 
## Models where empirical H_e is outside CI for Hotspot_chr1_window_1 :
##  [1] "s=0.075" "s=0.1"   "s=0.125" "s=0.15"  "s=0.2"   "s=0.3"   "s=0.4"  
##  [8] "s=0.5"   "s=0.6"   "s=0.7"   "s=0.8"
\end{verbatim}

\paragraph{6.4.3.2: Standard Error CI for Window
Length}\label{standard-error-ci-for-window-length}

The following section displays plot(s) comparing the average Window
length of the empirical ROH hotspot(s) with the corresponding standard
error-based confidence intervals for the different causative variant
windows.

\begin{verbatim}
## Models where empirical Length is within CI for Hotspot_chr1_window_1 :
## character(0)
## 
## Models where empirical Length is outside CI for Hotspot_chr1_window_1 :
##  [1] "s=0.075" "s=0.1"   "s=0.125" "s=0.15"  "s=0.2"   "s=0.3"   "s=0.4"  
##  [8] "s=0.5"   "s=0.6"   "s=0.7"   "s=0.8"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.3.2: Standard Error CI for Window Length-1.pdf}
\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.3.2: Standard Error CI for Window Length-2.pdf}

\begin{verbatim}
## Models where empirical Length is within CI for Hotspot_chr28_window_1 :
## [1] "s=0.15" "s=0.2"  "s=0.3" 
## 
## Models where empirical Length is outside CI for Hotspot_chr28_window_1 :
## [1] "s=0.075" "s=0.1"   "s=0.125" "s=0.4"   "s=0.5"   "s=0.6"   "s=0.7"  
## [8] "s=0.8"
\end{verbatim}

\paragraph{6.4.3.3: Standard Error CI for ROH
Freq}\label{standard-error-ci-for-roh-freq}

The following section displays plot(s) comparing the average
ROH-frequency of the empirical ROH hotspot(s) with the corresponding
standard error-based confidence intervals for the different causative
variant windows.

\begin{verbatim}
## Models where empirical ROH-freq is within CI for Hotspot_chr1_window_1 :
## character(0)
## 
## Models where empirical ROH-freq is outside CI for Hotspot_chr1_window_1 :
##  [1] "s=0.075" "s=0.1"   "s=0.125" "s=0.15"  "s=0.2"   "s=0.3"   "s=0.4"  
##  [8] "s=0.5"   "s=0.6"   "s=0.7"   "s=0.8"
\end{verbatim}

\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.3.3: Standard Error CI for ROH Freq-1.pdf}
\includegraphics{Pipeline_results_Windows_OS_files/figure-latex/6.4.3.3: Standard Error CI for ROH Freq-2.pdf}

\begin{verbatim}
## Models where empirical ROH-freq is within CI for Hotspot_chr28_window_1 :
## [1] "s=0.075" "s=0.1"  
## 
## Models where empirical ROH-freq is outside CI for Hotspot_chr28_window_1 :
## [1] "s=0.125" "s=0.15"  "s=0.2"   "s=0.3"   "s=0.4"   "s=0.5"   "s=0.6"  
## [8] "s=0.7"   "s=0.8"
\end{verbatim}

\end{document}
